PatternPal/PatternPal.Core/Checks/NodeCheck.cs

```diff
@@ -0,0 +1,201 @@
+
+    // The entities matched by this check.
+    private readonly List< INode > _matchedEntities;
+
+    // The current sub-check being checked.
+    private ICheck ? _currentSubCheck;
+
+    /// <summary>
+    /// Gets a <see cref="Func{TResult}"/> which returns a <see cref="List{T}"/> of <see cref="IEntity"/>s matched by this <see cref="ICheck"/>.
+    /// </summary>
+    /// <returns>A <see cref="List{T}"/> of matched <see cref="IEntity"/>s.</returns>
+    internal Func< List< INode > > Result => () => _matchedEntities;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="NodeCheck{TNode}"/> class.
+    /// </summary>
+    /// <param name="priority"><see cref="Priority"/> of the check.</param>
+    /// <param name="checks">A list of sub-<see cref="ICheck"/>s that should be checked.</param>
+    protected NodeCheck(
+        Priority priority,
+        IEnumerable< ICheck > subChecks,
+        CheckCollectionKind kind = CheckCollectionKind.All)
+        : base(priority)
+    {
+        _subChecks = subChecks;
+        _kind = kind;
+        _matchedEntities = new List< INode >();
+    }
+
+    /// <inheritdoc />
+    public override ICheckResult Check(
+        RecognizerContext ctx,
+        INode node)
+    {
+        // Verify that the node can be handled by this check.
+        TNode castNode = CheckHelper.ConvertNodeElseThrow< TNode >(node);
+
+        IList< ICheckResult > subCheckResults = new List< ICheckResult >();
+        foreach (ICheck subCheck in _subChecks)
+        {
+            _currentSubCheck = subCheck;
+            switch (subCheck)
+            {
+                case ClassCheck classCheck:
+                {
+                    subCheckResults.Add(
+                        classCheck.Check(
+                            ctx,
+                            castNode));
+                    break;
+                }
+                case CheckCollection checkCollection:
+                {
+                    subCheckResults.Add(
+                        checkCollection.Check(
+                            ctx,
+                            castNode));
+                    break;
+                }
+                case ModifierCheck modifierCheck:
+                {
+                    subCheckResults.Add(
+                        modifierCheck.Check(
+                            ctx,
+                            castNode));
+                    break;
+                }
+                case MethodCheck methodCheck:
+                {
+                    IEntity entity = CheckHelper.ConvertNodeElseThrow< IEntity >(node);
+                    foreach (IMethod method in entity.GetMethods())
+                    {
+                        subCheckResults.Add(
+                            methodCheck.Check(
+                                ctx,
+                                method));
+                    }
+                    break;
+                }
+                case FieldCheck fieldCheck:
```

Comment:
You have specified this for specifically a class. I would instead add a method GetFields which ClassCheck implements

Reply:
same for constructorcheck and propertycheck,

Reply:
I don't think the added complexity is worth it in this case.